internal/keyvault/keyvault.go

```diff
@@ -0,0 +1,179 @@
+package keyvault
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/keyvault/azcertificates"
+	"github.com/Azure/azure-sdk-for-go/sdk/keyvault/azkeys"
+	"github.com/Azure/azure-sdk-for-go/sdk/keyvault/azsecrets"
+	"github.com/Azure/notation-azure-kv/internal/crypto"
+)
+
+// clientAuthMode is the authorize mode used by plugin
+type clientAuthMethod string
+
+// set of auth errors
+var (
+	errMsgAuthorizerFromMI  = "authorized from Managed Identity failed, please ensure you have assigned identity to your resource"
+	errMsgAuthorizerFromCLI = "authorized from Azure CLI 2.0 failed, please ensure you have logged in using the 'az' command line tool"
+	errMsgUnknownAuthorizer = "unknown authorize method, please use a supported authorize method according to the document"
+)
+
+const (
+	// authMethodKey is the environment variable plugin used
+	authMethodKey = "AKV_AUTH_METHOD"
+	// authorizerFromMI auth akv from Managed Identity
+	// The auth order will be:
+	// 1. Client credentials
+	// 2. Client certificate
+	// 3. Username password
+	// 4. MSI
+	authorizerFromMI clientAuthMethod = "AKV_AUTH_FROM_MI"
+
+	// authorizerFromCLI auth akv from Azure cli 2.0
+	authorizerFromCLI clientAuthMethod = "AKV_AUTH_FROM_CLI"
+
+	// defaultAuthMethod is the default auth method if user doesn't provide an environment variable
+	// the default value will be authorizerFromCLI
+	defaultAuthMethod = authorizerFromCLI
+)
+
+// Certificate represents a Azure Certificate Vault.
+type Certificate struct {
+	keyClient    *azkeys.Client
+	certClient   *azcertificates.Client
+	secretClient *azsecrets.Client
+
+	name    string
+	version string
+}
+
+func NewCertificateFromID(id string) (*Certificate, error) {
+	u, err := url.Parse(id)
+	if err != nil {
+		return nil, fmt.Errorf("invalid certificate/key identifier: %q. the preferred schema is https://{vaultHost}/[certificates|keys]/{keyName}/{version}", id)
+	}
+
+	if u.Scheme != "https" {
+		return nil, fmt.Errorf("invalid certificate/key identifier: %q. the preferred schema is https://{vaultHost}/[certificates|keys]/{keyName}/{version}", id)
+	}
+
+	parts := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
+	if len(parts) != 3 || (parts[0] != "keys" && parts[0] != "certificates") {
+		return nil, fmt.Errorf("invalid certificate/key identifier: %q. the preferred schema is https://{vaultHost}/[certificates|keys]/{keyName}/{version}", id)
+	}
```

Comment:
The error message code block is duplicated. Can we refactor this block of code to a private function?

Reply:
It is also easier for unit tests.

Reply:
Refactored.